tests/scripts/generate_tls13_compat_tests.py

```diff
@@ -0,0 +1,486 @@
+import argparse
+import itertools
+
+# pylint: disable=useless-super-delegation
+
+CERTIFICATES = {
+    'ecdsa_secp256r1_sha256': (
+        'data_files/ecdsa_secp256r1.crt',
+        'data_files/ecdsa_secp256r1.key'),
+    'ecdsa_secp384r1_sha384': (
+        'data_files/ecdsa_secp384r1.crt',
+        'data_files/ecdsa_secp384r1.key'),
+    'ecdsa_secp521r1_sha512': (
+        'data_files/ecdsa_secp521r1.crt',
+        'data_files/ecdsa_secp521r1.key'),
+    'rsa_pss_rsae_sha256': (
+        'data_files/server2-sha256.crt', 'data_files/server2.key'
+    )
+}
+
+
+CAFILE = {
+    'ecdsa_secp256r1_sha256': 'data_files/test-ca2.crt',
+    'ecdsa_secp384r1_sha384': 'data_files/test-ca2.crt',
+    'ecdsa_secp521r1_sha512': 'data_files/test-ca2.crt',
+    'rsa_pss_rsae_sha256': 'data_files/test-ca_cat12.crt'
+}
+
+CIPHER_SUITE_IANA_VALUE = {
+    "TLS_AES_128_GCM_SHA256": 0x1301,
+    "TLS_AES_256_GCM_SHA384": 0x1302,
+    "TLS_CHACHA20_POLY1305_SHA256": 0x1303,
+    "TLS_AES_128_CCM_SHA256": 0x1304,
+    "TLS_AES_128_CCM_8_SHA256": 0x1305
+}
+
+SIG_ALG_IANA_VALUE = {
+    "ecdsa_secp256r1_sha256": 0x0403,
+    "ecdsa_secp384r1_sha384": 0x0503,
+    "ecdsa_secp521r1_sha512": 0x0603,
+    'rsa_pss_rsae_sha256': 0x0804,
+}
+
+NAMED_GROUP_IANA_VALUE = {
+    'secp256r1': 0x17,
+    'secp384r1': 0x18,
+    'secp521r1': 0x19,
+    'x25519': 0x1d,
+    'x448': 0x1e,
+}
+
+
+def remove_duplicates(seq):
+    seen = set()
+    seen_add = seen.add
+    return [x for x in seq if not (x in seen or seen_add(x))]
+
+
+class TLSProgram(metaclass=abc.ABCMeta):
+    """
+    Base class for generate server/client command.
+    """
+
+    def __init__(self, ciphersuite, signature_algorithm, named_group):
+        self._cipher = ciphersuite
+        self._sig_alg = signature_algorithm
+        self._named_group = named_group
+        self.add_ciphersuites(ciphersuite)
+        self.add_named_groups(named_group)
+        self.add_signature_algorithms(signature_algorithm)
+
+    @abc.abstractmethod
+    def add_ciphersuites(self, *ciphersuites):
+        pass
+
+    @abc.abstractmethod
+    def add_signature_algorithms(self, *signature_algorithms):
+        pass
+
+    @abc.abstractmethod
+    def add_named_groups(self, *named_groups):
+        pass
+
+    @abc.abstractmethod
+    def pre_checks(self):
+        return []
+
+    @abc.abstractmethod
+    def cmd(self):
+        pass
+
+    @abc.abstractmethod
+    def post_checks(self):
+        return []
+
+
+class OpenSSLServ(TLSProgram):
+    """
+    Generate test commands for OpenSSL server.
+    """
+    program = '$OPENSSL_NEXT'
+
+    def __init__(
+            self,
+            ciphersuite,
+            signature_algorithm,
+            named_group):
+        self.ciphersuites = []
+        self.named_groups = []
+        self.signature_algorithms = []
+        self.certificates = []
+        super().__init__(ciphersuite, signature_algorithm, named_group)
+
+    def add_ciphersuites(self, *ciphersuites):
```

Comment:
Shouldn't we test for duplication: adding a ciphersuite that is already in `self.ciphersuites` ? Ditto for the other `add_` functions.

Reply:
Refactor. Remove `add_*` from sub classes